### PR TITLE
Update the RNN cell API to be explicit about output_size.

### PR DESCRIPTION
**Pull request by qlzh727**
The existing contract of state_size[0] to be output size is bit
weird and might cause unnecessary contrain for supporting higher
dimension input/output/states. Adding explicit output_size to cell
so that the contract is not inferred by state_size anymore.

For backward compat reason, if a cell does not implement
output_size, the state_size[0] will be used as output_size.

The stackedRNNCells is also updated to by default return natural
order state_size as the cells' state, instead of reverse order.
For any user who rely on the states of the RNN, eg
RNN(return_state=True), please add the extra param
reverse_state_order=True while constructing the StackedRNNCells.

### PR Overview

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [x] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [x] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -54,36 +54,56 @@ def __init__(self, cells, **kwargs):
...
                                  'received cells:', cells)
         self.cells = cells
+        # reverse_state_order determines whether the state size will be in a
+        # reverse order of the cells' state. User might want to set this to True
+        # to keep the existing behavior. This is only useful when use
+        # RNN(return_state=True since the state will be returned as the same
+        # order of state_size.
+        self.reverse_state_order = kwargs.pop('reverse_state_order', False)
+        if self.reverse_state_order:
+            warnings.warn('reverse_state_order=True in StackedRNNCells will '
```
Comment (fchollet) [inline keras/layers/recurrent.py]: Please use backquotes around code keywords.

**Comment by qlzh727** [comment]
Seems that k.int_shape(x) returns None for the initial_state. From the code, it seems that int_shape() return None when its does not have keras_shape attribute.

**Comment by qlzh727** [comment]
Will take a look for the failed tests tomorrow.

**Comment by fchollet** [comment]
> From the code, it seems that int_shape() return None when its does not have keras_shape attribute.

Theano does not do static shape inference, so we have to rely on our own shape inference (keras_shape). The fix may be to add a line to propagate keras_shape where it goes missing. 

CNTK does do static shape inference. It seems the issue is different in this case.

**Comment by qlzh727** [comment]
The CNTK error is very weird, which does not make sense at all. The error message indicate that somehow the input shape does not match to the cell state. Is there anything special about the CNTK backend? 

>       model.train_on_batch(np.zeros((6, 5, 5)), np.zeros((6, 32)))
tests/keras/layers/recurrent_test.py:506: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
keras/engine/training.py:1223: in train_on_batch
    outputs = self.train_function(ins)
keras/backend/cntk_backend.py:1991: in __call__
    output_values = self.metrics_func.eval(input_dict, as_numpy=False)
../../../miniconda/envs/test-environment/lib/python2.7/site-packages/cntk/ops/functions.py:708: in eval
    _, output_map = self.forward(arguments, outputs, device=device, as_numpy=as_numpy)
../../../miniconda/envs/test-environment/lib/python2.7/site-packages/cntk/internal/swig_helper.py:69: in wrapper
    result = f(*args, **kwds)
../../../miniconda/envs/test-environment/lib/python2.7/site-packages/cntk/ops/functions.py:842: in forward
    keep_for_backward)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = Composite(Combine): Input('input_44', [#], [* x 5]), Input('rnn_4_target', [#]...rnn_4_sample_weights', [#], []) -> Output('ElementTimes36594_Output_0', [], [])
args = ({Input('rnn_4_target', [#], [*]): Value([6 x 32], CPU), Input('rnn_4_sample_weights', [#], []): Value([6], CPU), Input('input_44', [#], [* x 5]): Value([6 x 5 x 5], CPU)}, {Output('ElementTimes36594_Output_0', [], []): None}, CPU, set([]))
    def _forward(self, *args):
>       return _cntk_py.Function__forward(self, *args)
E       ValueError: Node 'Minus36542' (Minus operation): Input dimensions [8] and [32] are not compatible.

**Comment by fchollet** [comment]
@souptc could you please assist with the strange CNTK error? Thank you.

**Comment by souptc** [comment]
Spend some time to debugging, two issues here:
1. In cntk_backend.py, rnn function, I made a mistake when setup the output and states. @qlzh727 , do you mind to change the [this line](url) to
`new_output = n_s[-1]`
that should help you to pass most cases.

2. For the new test case 'test_inconsistent_output_state_size' you added, I think you have to disable it for cntk. My understanding is that cntk does not support multiple input/output variables, so we can't handle the case that input and state has different shape. The current "RNN" api implementation in cntk_backend is kind of compromise to Keras api, base on the assumption that the last states is the output.

**Comment by qlzh727** [comment]
Thanks for the suggestion. Update the cntk backend and test cases.

**Comment by qlzh727** [comment]
Test passed, @fchollet PTAL.

File: tests/keras/layers/recurrent_test.py [hunk 2]
```diff
@@ -907,5 +920,48 @@ def test_rnn_cell_identity_initializer(layer_class):
                           np.concatenate([np.identity(units)] * num_kernels, axis=1))
 
 
+@keras_test
+@pytest.mark.skipif((K.backend() in ['cntk']), reason='Not supported.')
```
Comment (fchollet) [inline tests/keras/layers/recurrent_test.py]: You can just use `if(K.backend() == 'cntk', `

File: keras/layers/recurrent.py [hunk 3]
```diff
@@ -54,36 +54,56 @@ def __init__(self, cells, **kwargs):
...
-        # in reverse order of the cell stack.
-        # This allows to preserve the requirement
-        # `stack.state_size[0] == output_dim`.
-        # e.g. states of a 2-layer LSTM would be
-        # `[h2, c2, h1, c1]`
+        # States are a flat list of the individual cell state size.
+        # e.g. states of a 2-layer LSTM would be `[h1, c1, h2, c2]`.
         # (assuming one LSTM has states [h, c])
+        # In the case of reverse_state_order=True, the state_size will be
+        # [h2, c2, h1, c1].
```
Comment (fchollet) [inline keras/layers/recurrent.py]: Nit: use backquotes around code keywords in comments (like you do above).

File: keras/layers/recurrent.py [hunk 4]
```diff
@@ -54,36 +54,56 @@ def __init__(self, cells, **kwargs):
                                  '`state_size` attribute. '
                                  'received cells:', cells)
         self.cells = cells
+        # reverse_state_order determines whether the state size will be in a
+        # reverse order of the cells' state. User might want to set this to True
+        # to keep the existing behavior. This is only useful when use
+        # RNN(return_state=True) since the state will be returned as the same
```
Comment (fchollet) [inline keras/layers/recurrent.py]: Nit: use backquotes around code keywords in comments.

File: tests/keras/layers/recurrent_test.py (hunk 2 above)
Comment (qlzh727) [reply tests/keras/layers/recurrent_test.py]: Done.

File: keras/layers/recurrent.py (hunk 3 above)
Comment (qlzh727) [reply keras/layers/recurrent.py]: Done.

File: keras/layers/recurrent.py (hunk 4 above)
Comment (qlzh727) [reply keras/layers/recurrent.py]: Done.

**Pull request merged**